### DISCUSSION TITLE
Pre-Commit Hook

### DISCUSSION THREAD

**Discussion by KelSolaar** [lycheeverse/lychee#1360] [General] [answered]
Hello,

I was wondering if it would make sense to have lychee available with [pre-commit](https://pre-commit.com). I would likely use it this way rather than with the Github Actions integration.

Cheers,

Thomas

**Comment by mre** [comment]
Hey @KelSolaar,
I don't mind adding support for it.

From what I can see, here are the necessary steps:
1. Fork this repo
2. Create a `.pre-commit-hooks.yaml` file as described [here](https://pre-commit.com/#new-hooks)
3. Test it by pointing `pre-commit` to the fork
4. Once it works, create a pull request.

If you have the time, feel free to look into it. I'd be happy to merge the final `.pre-commit-hooks.yaml` file so that others can profit from it as well. 👍 

Since this is technically not an issue with lychee, I will move the issue to GitHub discussions for the moment to keep the issue tracker clean.

**Comment by MattTimms** [comment, marked as answer]
Hi all,

I too have been interested in adding pre-commit support to lychee, so that I can check links on my site.  
 
I looked into the instructions from pre-commit's [new hooks docs](https://pre-commit.com/#new-hooks), however a `.pre-commit-hooks.yaml` with `language: rust` ([docs](https://pre-commit.com/#rust)) complains about finding a virtual manifest rather than a package manifest at root-level - seeing as Lychee's root-level `Cargo.toml` describes a workspace of packages.

I turned to `language: docker` ([docs](https://pre-commit.com/#docker)) where, after some minor hiccups (#1393 & #1394) I could get it working as expected. However, local docker build takes ~4min first-time on my machine & I bet downloading a prebuilt image at ~40MB compressed is preferably waaay faster.

So, I've come to a `.pre-commit-hook.yaml` that looks like this:
```yaml
  entry: lycheeverse/lychee:0.14
  language: docker_image
```

This would fetch the public lychee docker image & use it for subsequent pre-commit runs.

I'll create a pull request ([here](https://github.com/lycheeverse/lychee/pull/1396)) with this & an update to the readme, so that I can get some feedback. With the `.pre-commit-config.yaml` in place, pre-commit can be tested with `pre-commit run --verbose` & additionally with `--all-files`. 

Note that this solution would require updating that `entry:` field's docker tag with each new release of Lychee.
